scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -0,0 +1,157 @@
+            Literal.Const(e)
+          case (e: Execution.WriteExecution[a], _) =>
+            Literal.Const(e)
+          case (e: Execution.GetCounters[a], f) =>
+            widen(Literal.Unary[Execution, a, (a, ExecutionCounters)](f(e.prev), Execution.GetCounters(_: Execution[a])))
+          case (e: Execution.ResetCounters[a], f) =>
+            Literal.Unary(f(e.prev), Execution.ResetCounters(_: Execution[a]))
+          case (e: Execution.WithNewCache[a], f) =>
+            Literal.Unary(f(e.prev), Execution.WithNewCache(_: Execution[a]))
+          case (e: Execution.TransformedConfig[a], f) =>
+            Literal.Unary(f(e.prev), Execution.TransformedConfig(_: Execution[a], e.fn))
+          case (e: Execution.OnComplete[a], f) =>
+            Literal.Unary(f(e.prev), Execution.OnComplete(_: Execution[a], e.fn))
+          case (e: Execution.RecoverWith[a], f) =>
+            Literal.Unary(f(e.prev), Execution.RecoverWith(_: Execution[a], e.fn))
+          case (e: Execution.Mapped[a, b], f) =>
+            Literal.Unary(f(e.prev), Execution.Mapped(_: Execution[a], e.fn))
+          case (e: Execution.FlatMapped[a, b], f) =>
+            Literal.Unary(f(e.prev), Execution.FlatMapped(_: Execution[a], e.fn))
+          case (e: Execution.Zipped[a, b], f) =>
+            Literal.Binary(f(e.one), f(e.two), Execution.Zipped(_: Execution[a], _: Execution[b]))
+        }
+      }
+    )
+
+  object ZipWrite extends PartialRule[Execution] {
+    type InputType = (Config, Mode, Execution.Writer, ConcurrentExecutionContext)
+    type ResFun[T] =  InputType => Future[T]
+
+    case class NewFn[T, U](left: ResFun[T], right: ResFun[U]) extends (InputType => Future[(T, U)]) {
+      override def apply(tup: InputType): Future[(T, U)] =
+        Execution.failFastZip(left(tup), right(tup))(tup._4)
+    }
+
+    override def applyWhere[T](on: Dag[Execution]) = {
+      /*
+      * run this and that in parallel, without any dependency. This will
+      * be done in a single cascading flow if possible.
+      *
+      * If both sides are write executions then merge them
+      */
+      case Execution.Zipped(Execution.WriteExecution(oH, oT, oR), Execution.WriteExecution(tH, tT, tR)) =>
+        Execution.WriteExecution(oH, tH :: tT ::: oT, NewFn(oR, tR))
+    }
+  }
+
+  object ZipMap extends PartialRule[Execution] {
+    case class MapLeft[S, T, B](fn: S => B) extends (((S, T)) => (B, T)) {
+      override def apply(st: (S, T)): (B, T) = (fn(st._1), st._2)
+    }
+
+    case class MapRight[S, T, B](fn: T => B) extends (((S, T)) => (S, B)) {
+      override def apply(st: (S, T)): (S, B) = (st._1, fn(st._2))
+    }
+
+    override def applyWhere[T](on: Dag[Execution]) = {
+      case Execution.Zipped(Execution.Mapped(left, fn), right) =>
+        Execution.Zipped(left, right).map(MapLeft(fn))
+      case Execution.Zipped(left, Execution.Mapped(right, fn)) =>
+        Execution.Zipped(left, right).map(MapRight(fn))
```

Comment:
this is safe because it doesn't decrease parallelism.

Reply:
agree.